### PR TITLE
create a Token bi directional Translator

### LINKED ISSUE #7203: Token bi-directional translator

### PR DESCRIPTION
**Pull request by povolev15**
Implement the token object bi-directional convertor for migration

**Related issue(s)**:

Fixes #7203

### CODE REVIEW

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/merkle/MerkleToken.java [hunk 1]
```diff
@@ -176,8 +176,18 @@ public boolean equals(final Object o) {
                 && Objects.equals(this.symbol, that.symbol)
                 && Objects.equals(this.name, that.name)
                 && Objects.equals(this.memo, that.memo)
-                && Objects.equals(this.treasury, that.treasury)
-                && Objects.equals(this.autoRenewAccount, that.autoRenewAccount)
+                && Optional.ofNullable(this.treasury).stream()
```
Comment (jsync-swirlds) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/merkle/MerkleToken.java:179]: This seems massive overkill for a single object comparison.
Alternative:
```java
&& ((this.treasury == null && that.treasury == null)
        || (this.treasury != null && that.treasury != null
        && (this.treasury.num() == that.treasury.num())))
```
Similar for autoRenewAccount next.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/TokenStateTranslator.java [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+                .name(token.name())
+                .symbol(token.symbol())
+                .decimals(token.decimals())
+                .totalSupply(token.totalSupply())
+                .treasuryAccountNumber(token.treasury().num())
+                .lastUsedSerialNumber(token.getLastUsedSerialNumber())
+                .deleted(token.isDeleted())
+                .tokenType(fromMerkleType(token.tokenType()))
+                .supplyType(fromMerkleSupplyType(token.supplyType()))
+                .autoRenewAccountNumber(Optional.ofNullable(token.autoRenewAccount())
```
Comment (jsync-swirlds) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/TokenStateTranslator.java:62]: This would be simpler (and avoid creating 5 objects) as follows:
```java
                .autoRenewAccountNumber(token.autoRenewAccount() != null
                        ? token.autoRenewAccount().num()
                        : 0L)
```

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java [hunk 3]
```diff
@@ -50,6 +50,10 @@
 import com.hedera.pbj.runtime.io.stream.ReadableStreamingData;
 import com.hedera.pbj.runtime.io.stream.WritableStreamingData;
 import com.hederahashgraph.api.proto.java.AccountID.AccountCase;
+import com.hederahashgraph.api.proto.java.FixedFee;
```
Comment (jsync-swirlds) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java:53]: I believe that we have, in this class, been trying to always import the hapi version, and fully qualify only the hederahashgraph version of protobuf classes.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java [hunk 4]
```diff
@@ -1359,6 +1363,54 @@ public static com.hederahashgraph.api.proto.java.FileID fromPbj(FileID someFileI
...
+    public static FractionalFee fromPbj(com.hedera.hapi.node.transaction.FractionalFee fractionalFee) {
+        var builder = com.hederahashgraph.api.proto.java.FractionalFee.newBuilder();
+        builder.setFractionalAmount(fromPbj(fractionalFee.fractionalAmount()));
+        builder.setMinimumAmount(fractionalFee.minimumAmount());
+        builder.setMaximumAmount(fractionalFee.maximumAmount());
+        builder.setNetOfTransfers(fractionalFee.netOfTransfers());
+        return builder.build();
+    }
+
+    public static FixedFee fromPbj(com.hedera.hapi.node.transaction.FixedFee fixedFee) {
```
Comment (jsync-swirlds) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java:1405]: this is missing `@NonNull` annotation on return, as well as `@Nullable` on the parameter.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java [hunk 5]
```diff
@@ -1359,6 +1363,54 @@ public static com.hederahashgraph.api.proto.java.FileID fromPbj(FileID someFileI
...
+    }
+
+    public static Fraction fromPbj(com.hedera.hapi.node.base.Fraction fraction) {
+        var builder = com.hederahashgraph.api.proto.java.Fraction.newBuilder();
+        builder.setNumerator(fraction.numerator());
+        builder.setDenominator(fraction.denominator());
+        return builder.build();
+    }
+
+    public static FractionalFee fromPbj(com.hedera.hapi.node.transaction.FractionalFee fractionalFee) {
```
Comment (jsync-swirlds) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java:1396]: Nitpick:
This and the three above are missing `@NonNull` annotation on return and the parameter.
The parameter should probably be final as well.

File: hedera-node/hedera-mono-service/src/test/java/com/hedera/test/serde/EqualityType.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.test.serde;
+
+public enum EqualityType {
```
Comment (mhess-swl) [inline hedera-node/hedera-mono-service/src/test/java/com/hedera/test/serde/EqualityType.java:19]: Some javadoc explaining how this enum is meant to be used would be helpful

Comment (povolev15) [reply hedera-node/hedera-mono-service/src/test/java/com/hedera/test/serde/EqualityType.java:19]: @tinker-michaelj can you add it please?

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/merkle/MerkleToken.java (hunk 1 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/merkle/MerkleToken.java:179]: done thanks

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/TokenStateTranslator.java (hunk 2 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/TokenStateTranslator.java:62]: done thanks

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java (hunk 3 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java:53]: fixed it thanks

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java (hunk 4 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java:1405]: done thanks

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java (hunk 5 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java:1396]: done thanks

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/merkle/MerkleToken.java
Comment (jsync-swirlds) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/merkle/MerkleToken.java:1]: Is this intended to be added to this PR?  Seems unrelated...

File: hedera-node/hedera-mono-service/src/test/java/com/hedera/test/serde/EqualityType.java (hunk 6 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/test/java/com/hedera/test/serde/EqualityType.java:27]: @mhess-swl please check it up

File: hedera-node/hedera-mono-service/src/test/java/com/hedera/test/serde/SerializedForms.java [hunk 7]
```diff
@@ -152,7 +152,7 @@ private static <T extends FastCopyable> void assertSameCopySerialization(
     }
 
     private static void generateSerializedData() {
-        GENERATOR_MAPPING.get(SerializableSemVers.class).run();
+        GENERATOR_MAPPING.get(MerkleToken.class).run();
```
Comment (mhess-swl) [inline hedera-node/hedera-mono-service/src/test/java/com/hedera/test/serde/SerializedForms.java:155]: I don't think it's a big deal that this line changed, but want to check?

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/TokenStateTranslator.java [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+import com.hedera.node.app.service.token.ReadableTokenStore;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Translates between the legacy {@link com.hedera.node.app.service.mono.state.merkle.MerkleToken} and the {@link Token} and vise versa.
+ */
+public class TokenStateTranslator {
```
Comment (mhess-swl) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/TokenStateTranslator.java:41]: Nit: `public final class`

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/TokenStateTranslator.java [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+     * @param readableTokenStore the {@link com.hedera.node.app.service.token.ReadableTokenStore} to use to retrieve the token
+     * @return the {@link com.hedera.node.app.service.mono.state.merkle.MerkleToken} corresponding to the tokenId
+     */
+    public static com.hedera.node.app.service.mono.state.merkle.MerkleToken merkleTokenFromToken(
+            @NonNull TokenID tokenId, @NonNull ReadableTokenStore readableTokenStore) {
+        requireNonNull(tokenId);
+        requireNonNull(readableTokenStore);
+        final var optionalToken = readableTokenStore.get(tokenId);
+        if (optionalToken == null) {
+            throw new IllegalArgumentException("Account not found");
```
Comment (mhess-swl) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/TokenStateTranslator.java:141]:
```suggestion
            throw new IllegalArgumentException("Token not found");
```

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/TokenStateTranslator.java (hunk 8 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/TokenStateTranslator.java:41]: done thanks

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/TokenStateTranslator.java (hunk 9 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/TokenStateTranslator.java:141]: done thanks

File: hedera-node/hedera-mono-service/src/test/java/com/hedera/test/serde/SerializedForms.java (hunk 7 above)
Comment (povolev15) [reply hedera-node/hedera-mono-service/src/test/java/com/hedera/test/serde/SerializedForms.java:155]: As I understand it doesn't matter. @tinker-michaelj this is just generate the data and it is fine

**Review by mhess-swl — APPROVED** [review]
LGTM! Ty!

**Review by Neeharika-Sompalli — APPROVED** [review]

**Review by JeffreyDallas — APPROVED** [review]

**Review by qnswirlds — APPROVED** [review]
